### PR TITLE
ClientModel: Remove OptionalClientResult<T> in favor of using nullable annotations

### LINKED ISSUE #40568: ClientModel: Evaluate the feasibility of using nullable annotations and removing OptionalOutputMessage<T> type
Per discussion with BCL team, consider the following API change:

- Remove `OptionalClientResult<T>`
- Change client APIs to return `ClientResult<T>` when it is guaranteed to return a model or throw if no model is returned, and `ClientResult<T?>` when it is valid that no model could be returned, e.g. `GetXxIfExists` APIs.

The benefits of this approach include:
  - The compiler will give them nullability reference type warnings or nullable reference type warnings if they try and dereference without a check 1st and the check is how they determine whether it has a value or not.

The considerations to determine whether we can use the design proposal include:
  - What is the hierarchy across Azure.Core 2.0?
    - Previously documented:
      - [Krzysztof's gist](https://gist.github.com/KrzysztofCwalina/2473c1d1423b4f131070761999aa64e3)
      - [Anne's gist](https://gist.github.com/annelo-msft/58277d8bce442519d4191fe72ffab10b)
  - Will it work with structs?
    - Must support `Response<bool>` for a HEAD request, e.g.
  - How would we need to change client settings, which don't currently use <Nullable>enable</Nullable>
  - How does it work with valid JSON value of JSON `null`?

### PR DESCRIPTION
**Pull request by annelo-msft**
Closes https://github.com/Azure/azure-sdk-for-net/issues/40568

Per discussion in https://github.com/Azure/azure-sdk-for-net/issues/40568, this PR removes `OptionalClientResult<T>` from System.ClientModel in favor of using nullable annotations to indicate that a client's service method may or may not return a deserialized model in the returned ClientResult.

Demo client implementations are provided in the investigation PR: https://github.com/Azure/azure-sdk-for-net/pull/41499

With this change, ClientModel-based clients will return `ClientResult<T>` when a service method is guaranteed to return a model or throw if no model can be returned, and `ClientResult<T?>` when it is valid that no model could be returned, e.g. in a `GetIfExists` API.  Generated clients will need to set `<Nullable>enabled</Nullable>` in their .csproj files.

### CODE REVIEW

File: sdk/core/System.ClientModel/tests/TestFramework/Mocks/MockJsonModel.cs [hunk 1]
```diff
@@ -10,15 +10,15 @@
 
 namespace ClientModel.Tests.Mocks;
 
-public class MockPersistableJsonModel : IJsonModel<MockPersistableJsonModel>
+public class MockJsonModel : IJsonModel<MockJsonModel>
```
Comment (annelo-msft) [inline sdk/core/System.ClientModel/tests/TestFramework/Mocks/MockJsonModel.cs:13]: Just renamed this type to better match the interface name.

File: sdk/core/System.ClientModel/tests/Convenience/ClientResultTests.cs [hunk 2]
```diff
@@ -38,46 +38,38 @@ public void GetRawResponseReturnsResponse()
...
     public void CanCreateOptionalClientResultFromBool()
     {
         // This tests simulates creation of the result returned from a HEAD request.
 
         PipelineResponse response = new MockPipelineResponse(200);
-        OptionalClientResult<bool> result = ClientResult.FromOptionalValue(true, response);
+        ClientResult<bool?> result = ClientResult.FromOptionalValue<bool?>(true, response);
 
         Assert.IsTrue(result.Value);
-        Assert.IsTrue(result.HasValue);
```
Comment (KrzysztofCwalina) [inline sdk/core/System.ClientModel/tests/Convenience/ClientResultTests.cs:62]: Can you call Assert.IsTrue(result.Value.HasValue)?

**Review by KrzysztofCwalina — APPROVED** [review]

File: sdk/core/System.ClientModel/tests/Convenience/ClientResultTests.cs (hunk 2 above)
Comment (annelo-msft) [reply sdk/core/System.ClientModel/tests/Convenience/ClientResultTests.cs:62]: Good catch!  Yes, will fix

**Comment by annelo-msft** [comment]
Filed https://github.com/Azure/azure-sdk-for-net/issues/41554 to track follow up to this PR.

File: sdk/core/System.ClientModel/tests/Convenience/ClientResultTests.cs (hunk 2 above)
Comment (annelo-msft) [reply sdk/core/System.ClientModel/tests/Convenience/ClientResultTests.cs:62]: Ah, I see, not a bug, but an additional test case.  Yes, will add

**Review by KrzysztofCwalina — APPROVED** [review]

**Pull request merged**